network/network/src/main/java/bisq/network/p2p/common/ResponseHandlerDelegate.java

```diff
@@ -0,0 +1,111 @@
+    private final Map<String, RequestFuture<T, R>> requestFuturesByConnectionId = new ConcurrentHashMap<>();
+
+    ResponseHandlerDelegate(Node node, long timeout, Class<R> responseClass) {
+        this.node = node;
+        this.timeout = timeout;
+        this.responseClass = responseClass;
+    }
+
+    public Optional<R> resolveResponse(EnvelopePayloadMessage message) {
+        return responseClass.isInstance(message)
+                ? Optional.of(responseClass.cast(message))
+                : Optional.empty();
+    }
+
+
+    // If we get called on a connection we have already assigned a request for, we ignore the new request and return
+    // the future from the pending request.
+    public CompletableFuture<R> request(Connection connection, T request) {
+        return requestFuturesByConnectionId.compute(connection.getId(), (k, existing) -> {
+            if (existing != null && !existing.isDone()) {
+                log.warn("[{}] Reusing pending request future for {}", request.getClass().getSimpleName(), connection.getPeerAddress());
+                return existing;
+            }
+            RequestFuture<T, R> requestFuture = new RequestFuture<>(node, connection, request);
+            requestFuture.orTimeout(timeout, TimeUnit.MILLISECONDS)
+                    .whenComplete((response, throwable) -> {
+                        requestFuturesByConnectionId.remove(k);
+                        if (throwable instanceof TimeoutException) {
+                            log.warn("[{}] Request to {} timed out after {} ms", request.getClass().getSimpleName(), connection.getPeerAddress(), timeout);
+                        } else if (throwable != null) {
+                            log.warn("[{}] Request to {} failed: {}", request.getClass().getSimpleName(), connection.getPeerAddress(), ExceptionUtil.getRootCauseMessage(throwable));
+                        } else {
+                            log.debug("[{}] Request to {} completed", request.getClass().getSimpleName(), connection.getPeerAddress());
+                        }
+                    });
+            return requestFuture;
+        });
+    }
+
+    protected void processResponse(Connection connection, R response) {
+        Optional.ofNullable(requestFuturesByConnectionId.get(connection.getId())).ifPresent(handler -> handler.handleResponse(response));
+    }
+
+    public void processOnDisconnect(Connection connection, CloseReason closeReason) {
+        requestFuturesByConnectionId.computeIfPresent(connection.getId(), (key, handler) -> {
+            if (!handler.isDone()) {
+                handler.cancel(true);
+            }
+            return null; // removes entry
+        });
+    }
+
+    public Map<String, RequestFuture<T, R>> getRequestFuturesByConnectionId(){
+        return requestFuturesByConnectionId;
+    }
+
+    /* --------------------------------------------------------------------- */
+    // LifecycleHandler implementation
```

Comment:
Small style nit: We usually place the initialize and shutdown after constructor.